### ISSUE TITLE
Include CallActivityAsync() with void return type in IDurableOrchestrationContext to enable testing with mock

### ISSUE THREAD

**Issue by anthonychu** [Azure/azure-functions-durable-extension#1224]
Copied from [UserVoice](https://feedback.azure.com/forums/355860-azure-functions/suggestions/39072883-inlcude-callactivityasync-with-void-return-type):

> We tried to unit test orchestrator with a mock IDurableOrchestrationContext, but CallActivityAsync() with void return type is in DurableContextExtensions rather the interface. The workaround we did is to return a dummy value from the Function, and test it as CallActivityAsync<TResult>() from the IDurableOrchestrationContext.

**Comment by cgillum** [comment]
Adding @ConnorMcMahon since this relates to some of our previous conversations about the problems with C# interfaces.

**Comment by olitomlinson** [comment]
This is currently blocking me from upgrading from V1 to V2 DF :(

I shouldn't have change my method signatures to work around this issue (as was proposed in the original UserVoice comment)

**Comment by ConnorMcMahon** [comment]
@olitomlinson, to clarify, I think you can get around this in the meantime by only mocking the core `CallActivityAsync()` method. Your code inside of your orchestrator should still be able to use the `void` return type method signature. That being said, you shouldn't have to know which interface method is being called by the extension methods in order to mock the behavior of the extension methods. This has been a confusing issue for many customers.

I'm wondering if the best workaround for this is to stop using extension methods and just add all of the methods directly on the interface. Since we only have one implementation of each interface, and customers should be using Moq to test the interfaces, I am questioning the utility in using extension methods instead of just populating each method directly on the interface. 

If we went with this approach, user's could mock any of our method signatures, instead of just the "core" interface methods.

@sebastianburckhardt thoughts?

**Comment by olitomlinson** [comment]
@ConnorMcMahon In my code, I don't set-up (mock a response) to the method being called. All I need to do is `verify` the method call, but that doesn't appear to work.

``` 
Message: 
    System.NotSupportedException : Unsupported expression: c => c.CallActivityAsync("doTheThing", It.Is<Foo>(r => r.Result.Result == "true" && r.Result.Reason == "x"))
    Extension methods (here: DurableContextExtensions.CallActivityAsync) may not be used in setup / verification expressions.
```

Can you advise if there is another way to do this? Cheers!

**Comment by ConnorMcMahon** [comment]
@olitomlinson 

I believe if you replace the lambda `c => c.CallActivityAsync("doTheThing", It.Is<Foo>(r => r.Result.Result == "true" && r.Result.Reason == "x")` with `c => c.CallActivityAsync<object>("doTheThing", It.Is<Foo>(r => r.Result.Result == "true" && r.Result.Reason == "x")` in your `Verify()` call, it should fix the issue.

This is still annoying from the feedback I have gotten from customers, so I am leaning towards moving all of our extension methods to explicit interface methods in 2.3.0. This would be a breaking change for customer tests written without the use of Moq, so we try to keep those in minor releases instead of patch releases, and clearly document the test changes.

**Comment by olitomlinson** [comment]
@ConnorMcMahon your suggestion worked perfectly. Many thanks!

**Comment by olitomlinson** [comment]
@ConnorMcMahon Is there a work around for mocking `DurableClient.CreateEntityProxy` ? I'm getting the same problem with this being an extension. Thanks in advance.

**Comment by ConnorMcMahon** [comment]
Unfortunately, that one is a bit trickier.

Most of our extension methods are just method overloads that are proxying a call to the actual interface method with the correct default values.

This one, however, is functionally a wrapper for some more complicated logic around our proxy creation. 

I will file a separate issue for that method. In the meantime, you could potentially work around it by using your own dependency injection, in which you have a proxy factory, and the production path uses our code path, and the test path uses your mocks.

**Comment by olitomlinson** [comment]
@ConnorMcMahon Sorry yes! I did mean `IDurableOrchestrationContext` .

Yeah, I'll wrap it with my own interface in the mean time. Cheers for the help!

**Comment by ravibha** [comment]
@ConnorMcMahon Is there a similar workaround for the Setup method? I am doing something like this in my code:
```
mockDurableOrchestrationContext.Setup(x => x.CallActivityAsync(FunctionName, It.IsAny<object>()))
                .Returns(
                    (string functionName, object input) =>
                    {
                        durableActivityContextMock.Setup(mock => mock.GetInput<object>())
                            .Returns(() => input as CustomEvent);

                        return activityFunction.RunAsync(durableActivityContextMock.Object);
                    });
```
But it throws an NotSupportedException when the test case runs.
```
System.NotSupportedException : Unsupported expression: x => x.CallActivityAsync(FunctionName, It.IsAny<object>())
    Extension methods (here: DurableContextExtensions.CallActivityAsync) may not be used in setup / verification expressions.
```

**Comment by mpaul31** [comment]
@ConnorMcMahon it appears this extension method is not mockable as well correct due to the EntityClientProxy usage from within?

`public static Task SignalEntityAsync<TEntityInterface>(this IDurableEntityClient client, string entityKey, Action<TEntityInterface> operation)`

Also, is there a reason why the durable clients are not able to be registered in the DI container?

**Comment by ConnorMcMahon** [comment]
This was largely fixed in v2.3.0 (with a few stragglers methods in v2.4.0) with our transition away from using extension methods.

@mpaul31, with our v2.4.0 release we now support Durable clients via DI.
